### PR TITLE
Support marking Services as firewall-unmanaged

### PR DESCRIPTION
**Pull request by timoreimann**
This change enables more direct control over how `NodePort` Services should be accessible through a new configuration annotation. One use case is when a `NodePort` is to be used for VPC-internal connectivity only, in which case no public access is desirable at all.

We also extract boolean annotation value parsing into a dedicated helper function and refactoring existing usages accordingly.

### CODE REVIEW

File: cloud-controller-manager/do/loadbalancers.go [hunk 1]
```diff
@@ -1131,48 +1122,30 @@ func getStickySessionsCookieTTL(service *v1.Service) (int, error) {
...
-	redirectHTTPToHTTPS, ok := service.Annotations[annDORedirectHTTPToHTTPS]
-	if !ok {
-		return false, nil
-	}
-
-	redirectHTTPToHTTPSBool, err := strconv.ParseBool(redirectHTTPToHTTPS)
+	redirectHTTPToHTTPS, _, err := getBool(service.Annotations, annDORedirectHTTPToHTTPS)
 	if err != nil {
-		return false, fmt.Errorf("failed to parse redirect HTTP-to-HTTPS flag %q from annotation %q: %s", redirectHTTPToHTTPS, annDORedirectHTTPToHTTPS, err)
+		return false, fmt.Errorf("failed to get HTTP-to-HTTPS configuration setting: %s", err)
```
Comment (MorrisLaw) [inline cloud-controller-manager/do/loadbalancers.go:1127]: Why the change from `parse` to `get`?

**Review by MorrisLaw — COMMENTED** [review]
Left some comments/questions, it’s looking good!

File: cloud-controller-manager/do/firewall_controller.go [hunk 2]
```diff
@@ -395,6 +410,18 @@ func (fm *firewallManager) createReconciledFirewallRequest(serviceList []*v1.Ser
...
+//isManaged returns if the given Service should be firewall-managed based on the
+//configuration annotation. An omitted annotation applies the default behavior
+//of managing firewall rules for the Service.
+func isManaged(service *v1.Service) (bool, error) {
+	val, found, err := getBool(service.Annotations, annotationDOFirewallManaged)
+	if err != nil {
+		return false, err
+	}
+
+	return !found || val, nil
```
[the remark this answers is not in the document]
Comment (timoreimann) [reply cloud-controller-manager/do/firewall_controller.go:422]: `val` would be `false` if the value was not found. However, we want to consider a Service to be managed if the annotation was _not_ found (i.e., was absent). I tried to make this clear in the GoDoc.

File: cloud-controller-manager/do/loadbalancers.go (hunk 1 above)
Comment (timoreimann) [reply cloud-controller-manager/do/loadbalancers.go:1127]: My thinking here is that parsing is just one aspect of getting a parameter: the value needs to be retrieved from the annotation map and then converted (with the latter being what I considered the essential parsing step).

Before this change, parsing was directly implemented by `getRedirectHTTPToHTTPS`, but now it's just an implementation detail of a more abstract function (`getBool`). Maintaining the wording to me means we'd be leaking implementation details into this function.

I guess you could make an argument that we could return `err` directly without wrapping. The rationale we do so for now is that it presumably makes the error description more readable by not just printing a rather technical sounding annotation name (`service.beta.kubernetes.io/do-loadbalancer-redirect-http-to-https`) but also a descriptive text (_HTTP-to-HTTPS_). We could discuss whether that's worth keeping, though I'd rather do that in a separate PR that updates all descriptions in one go.

File: cloud-controller-manager/do/annotations.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	if !ok {
+		return false, false, nil
+	}
+
+	parsed, err := strconv.ParseBool(value)
+	if err != nil {
+		return false, false, fmt.Errorf("cannot convert value %q for annotation %q to bool: %s", value, key, err)
+	}
+
+	return parsed, true, nil
```
[the remark this answers is not in the document]
Comment (timoreimann) [reply cloud-controller-manager/do/annotations.go:35]: See the other comment on why I think we need to include the `found` indicator.

File: cloud-controller-manager/do/firewall_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (MorrisLaw) [reply cloud-controller-manager/do/firewall_controller.go:422]: Ok, thank you for confirming that for me 😄

**Review by MorrisLaw — APPROVED** [review]
Looks good to me 👍
